catroid/src/main/java/org/catrobat/catroid/transfers/project/ProjectUpload.kt

```diff
@@ -0,0 +1,133 @@
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.util.Locale
+
+typealias UploadProjectSuccessCallback = (projectId: Int) -> Unit
+typealias UploadProjectErrorCallback = (errorCode: Int, errorMessage: String) -> Unit
+
+class ProjectUpload(
+    private val projectDirectory: File,
+    private val projectName: String,
+    private val projectDescription: String,
+    private val userEmail: String,
+    private val sceneNames: Array<String>?,
+    private val archiveDirectory: File,
+    private val zipArchiver: ZipArchiver,
+    private val screenshotLoader: ProjectAndSceneScreenshotLoader,
+    private val sharedPreferences: SharedPreferences,
+    private val serverCalls: ServerCalls
+) {
+
+    fun start(
+        successCallback: UploadProjectSuccessCallback,
+        errorCallback: UploadProjectErrorCallback
+    ) {
+        val projectArchive = zipProjectToArchive(projectDirectory, archiveDirectory)
+        if (projectArchive == null) {
+            errorCallback(UPLOAD_ZIP_ERROR, UPLOAD_ZIP_ERROR_MESSAGE)
+            return
+        }
+
+        val projectUploadData = createUploadData(projectArchive)
+
+        scaleSceneScreenshots(projectName, sceneNames)
+
+        serverCalls.uploadProject(
+            projectUploadData,
+            { projectId, successUsername, successToken ->
+                sharedPreferences.edit()
+                    .putString(Constants.TOKEN, successToken)
+                    .putString(Constants.USERNAME, successUsername)
+                    .commit()
+
+                successCallback(projectId)
+                projectArchive.delete()
+            },
+            { errorCode, errorMessage ->
+                errorCallback(
+                    errorCode,
+                    errorMessage
+                )
+            }
+        )
+    }
+
+    private fun createUploadData(projectArchive: File): ProjectUploadData {
+        val token = sharedPreferences.getString(Constants.TOKEN, Constants.NO_TOKEN)
+        val username = sharedPreferences.getString(Constants.USERNAME, Constants.NO_USERNAME)
+
+        return ProjectUploadData(
+            projectName = projectName,
+            projectDescription = projectDescription,
+            projectArchive = projectArchive,
+            userEmail = userEmail,
+            language = Locale.getDefault().language,
+            token = token,
+            username = username
+        )
+    }
+
+    private fun scaleSceneScreenshots(projectName: String, sceneNames: Array<String>?) {
+        sceneNames?.mapNotNull { screenshotLoader.getScreenshotFile(projectName, it, false) }
+            ?.filter { it.exists() && it.length() > 0 }
+            ?.forEach {
+                try {
+                    ImageEditing.scaleImageFile(it, UPLOAD_IMAGE_SCALE_WIDTH, UPLOAD_IMAGE_SCALE_HEIGHT)
+                } catch (ex: FileNotFoundException) {
+                    Log.e(TAG, Log.getStackTraceString(ex))
```

Comment:
share has its default screenshot thingy going on if no screenshot is contained in the uploaded file.
this can happen, eg. if one manually deletes the screenshot file, OR uploads a project before running the stage at least once.

if we want to incorporate a default or error screenshot, we should add it to every project on creation, and replace it once the stage is run and a real one is created.

Reply:
Additionally, I don't think this is within the scope of the ticket, so if we really do need a default screenshot, I'd prefer having a separate ticket for this.